### PR TITLE
Update Promise type annotations

### PR DESCRIPTION
**Pull request by icholy**
~~Here's a start for~~ https://github.com/AnalyticalGraphicsInc/cesium/issues/2730

### CODE REVIEW

File: Apps/Sandcastle/gallery/Custom DataSource.html [hunk 1]
```diff
@@ -189,7 +189,7 @@
 /**
  * Asynchronously loads the GeoJSON at the provided url, replacing any existing data.
  * @param {Object} url The url to be processed.
- * @returns {Promise} a promise that will resolve when the GeoJSON is loaded.
+ * @returns {Promise.<Any>} a promise that will resolve when the GeoJSON is loaded.
```
Comment (icholy) [inline Apps/Sandcastle/gallery/Custom DataSource.html]: Not sure what to use here `Any` or `Object`.

Comment (icholy) [inline Apps/Sandcastle/gallery/Custom DataSource.html]: Here's a reference of using `Any` https://github.com/icholy/cesium/blob/bd015ae51efe0b1bf8dcfe33df8e17a9b3bdc211/Source/Core/IndexDatatype.js#L98

**Comment by mramato** [comment]
Thanks.  I'll take a closer look at this soon. It looks like jsDoc added official support for Promise annotation in 3.3.0 (we are on beta8 but can probably upgrade easily), so I don't see why we shouldn't take advantage of them.

One thing I did notice is that many of your <Any> annotations appear to be incorrect (at first glance).  For example, loadImage always resolves to an `Image` instance, so I imagine you would specify that instead of Any.

**Comment by icholy** [comment]
@mramato its still a WIP, ill ping you when it's ready for review. I wasn't aware of the promise support in jsdoc, I'll check it out.

File: Source/Core/EarthOrientationParameters.js [hunk 2]
```diff
@@ -135,7 +135,7 @@ define([
      * Gets a promise that, when resolved, indicates that the EOP data has been loaded and is
      * ready to use.
      *
-     * @returns {Promise} The promise.
+     * @returns {Promise.<undefined>} The promise.
```
Comment (icholy) [inline Source/Core/EarthOrientationParameters.js]: Not sure about this one

File: Source/Core/EarthOrientationParameters.js [hunk 3]
```diff
@@ -135,7 +135,7 @@ define([
      * Gets a promise that, when resolved, indicates that the EOP data has been loaded and is
      * ready to use.
      *
-     * @returns {Promise} The promise.
+     * @returns {Promise.<Object>} The promise.
```
Comment (icholy) [inline Source/Core/EarthOrientationParameters.js]: Should this be `EarthOrientationParameters` ?

File: Source/Core/Transforms.js [hunk 4]
```diff
@@ -523,7 +523,7 @@ define([
      * indicates that the preload has completed.
      *
      * @param {TimeInterval} timeInterval The interval to preload.
-     * @returns {Promise} A promise that, when resolved, indicates that the preload has completed
+     * @returns {Promise.<undefined>} A promise that, when resolved, indicates that the preload has completed
```
Comment (icholy) [inline Source/Core/Transforms.js]: Not sure about this one.

**Comment by icholy** [comment]
@mramato I think it's almost there. Take a look when you have a minute.

I can't find the jsdoc promise support you mentioned. I read through [this issue](https://github.com/jsdoc3/jsdoc/issues/509), but there doesn't seem to be anything implemented.

File: Source/Core/EarthOrientationParameters.js (hunk 3 above)
Comment (shunter) [inline Source/Core/EarthOrientationParameters.js]: Currently, this promise resolves to no value, i.e. `undefined`.  We use this pattern in a few places to indicate that some process has finished, but there is no obvious result value of the process.

This particular class is marked "private", though, and is considered undocumented, so I don't think we need to worry about it for now.

In general, perhaps we should perhaps always try to return some value from all promises, even when not really necessary.  Perhaps as you suggest this loading promise could resolve to the EarthOrientationParameters object, even though the caller must have already had a reference to that object in order to access the promise in the first place.

File: Source/Core/Transforms.js (hunk 4 above)
Comment (shunter) [inline Source/Core/Transforms.js]: Similar to above, the resolved value is not useful.  From examining the code, it's an array of array of undefined, but that shouldn't be documented.

Comment (icholy) [inline Source/Core/Transforms.js]: Ok, that's what I thought. Just wanted to double check.

**Comment by icholy** [comment]
Rebased & Squashed

**Comment by icholy** [comment]
Are there any issues with this this that I still need to address? Or is it just going to take some time to get to/review?

**Comment by mramato** [comment]
We just need to take the time for a closer look and review.

**Comment by pjcozzi** [comment]
What's the plan with this PR?

**Comment by icholy** [comment]
I need these changes to move forward with making TypeScript definition files.

**Comment by mramato** [comment]
The main problem is determining the impact on our generated documentation.  Last I checked, while this was a proposed syntax for jsDoc, it wasn't supported yet.  I need to check out the latest version and see if that's changed.  Even if they don't support it, we may be able to take this change (as long as the doc still compiled and doesn't drop information).

**Comment by icholy** [comment]
Here's what it looks like
doc

**Comment by mramato** [comment]
Thanks @icholy I'll take a final look at this tomorrow and assuming no issues, we can merge it in.  Thanks again.

**Comment by mramato** [comment]
It may have to change in the future when JSDoc gets official Promise support, but it looks like they may choose this syntax anyway.  I'll make a tweak in master to change `Promise.<Any>` to `Promise.<Object>` for consistency, but that's about it.

**Comment by pjcozzi** [comment]
@icholy thanks again for this contribution.  We appreciate all your pull requests.

**Pull request merged**